File: src/lexer/cursor.rs
use crate::lexer::errors::LexError;

pub(crate) struct Cursor<'cursor> {
    code: &'cursor [char],
    pub(crate) current: usize,
}

impl<'cursor> Cursor<'cursor> {
    pub(crate) fn new(code: &'cursor [char]) -> Self {
        Cursor { code, current: 0 }
    }

    pub(crate) fn peek(&self) -> Option<char> {
        if self.is_at_end() {
            None
        } else {
            Some(self.offset(0)?)
        }
    }

    pub(crate) fn is_at_end(&self) -> bool {
        self.current >= self.code.len()
    }

    pub(crate) fn offset(&self, offset: usize) -> Option<char> {
        let index = self.current + offset;
        if self.code.len() > index {
            Some(self.code[index])
        } else {
            None
        }
    }

    pub(crate) fn offset_result(&self, offset: usize) -> Result<char, LexError> {
        let index = self.current + offset;
        if self.code.len() > index {
            Ok(self.code[index])
        } else {
            Err(LexError::UnexpectedEof)
        }
    }
}

File: internal/lexer/cursor.go
package lexer

type Cursor struct {
    code   []rune
    current int
}

func NewCursor(code []rune) Cursor {
    return Cursor{code: code, current: 0}
}

func (c *Cursor) Peek() (rune, bool) {
    if c.IsAtEnd() { return 0, false }
    r, ok := c.Offset(0)
    if !ok { return 0, false }
    return r, true
}

func (c *Cursor) IsAtEnd() bool {
    return c.current >= len(c.code)
}

func (c *Cursor) Offset(offset int) (rune, bool) {
    index := c.current + offset
    if index < len(c.code) {
        return c.code[index], true
    }
    return 0, false
}

func (c *Cursor) OffsetResult(offset int) (rune, error) {
    index := c.current + offset
    if index < len(c.code) {
        return c.code[index], nil
    }
    return 0, UnexpectedEOF()
}


